### PR TITLE
[JENKINS-34739] Handle varargs

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-34739](https://issues.jenkins-ci.org/browse/JENKINS-34739)

Replaces the fix in #58 (and the follow-up in #65).

@reviewbybees and CC @magnusr

### CODE REVIEW

**Comment by magnusr** [comment]
👍

**Comment by kwhetstone** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java [hunk 1]
```diff
@@ -43,7 +44,27 @@
...
-    private static boolean matches(@Nonnull Class<?>[] parameterTypes, @Nonnull Object[] parameters) {
+    private static boolean matches(@Nonnull Class<?>[] parameterTypes, @Nonnull Object[] parameters, boolean varargs) {
+        if (varargs) {
+            int fixedLen = parameterTypes.length - 1;
+            Class<?> componentType = parameterTypes[fixedLen].getComponentType();
+            assert componentType != null;
+            int arrayLength = parameters.length - fixedLen;
+            if (arrayLength >= 0) {
+                if (arrayLength == 1 && parameterTypes[fixedLen].isInstance(parameters[fixedLen])) {
+                    // not a varargs call
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: 🐛  Rather than redefining an argument in a fairly complex code block, let's move this to a separate method returning the new parameters (hard to follow otherwise).

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java [hunk 2]
```diff
@@ -43,7 +44,27 @@
...
+                } else {
+                    Object array = Array.newInstance(componentType, arrayLength);
+                    for (int i = 0; i < arrayLength; i++) {
+                        Array.set(array, i, parameters[fixedLen + i]);
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: I assume we're doing the set operation rather than arraycopy to handle primitives and boxing/unboxing?

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java [hunk 3]
```diff
@@ -140,28 +161,10 @@ private static boolean isInstancePrimitive(@Nonnull Class<?> type, @Nonnull Obje
         return findMatchingMethod(receiver, method, args);
     }
 
-    private static boolean isEnumInitializer(Method m) {
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: Er, why did we remove this?

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: No, could probably have used `arraycopy`.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 3 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: Because is has been superseded by the more general fix.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 2 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: 🐜 then I think this could be done more simply with one of the Arrays.copyOfRange methods?  Ditto the other one too, I think, reducing this to three simple lines.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: Never mind, you _do_ get an `ArrayStoreException`. Needs test coverage.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 1 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: To capture reasoning discussed OoB for making this a bug and not a nit:  

> Redefining inputs introduces unexpected behavior that others may not catch without reading the function in entirety and understanding it deeply. 

Also 🐜  that we can remove the initial if case, since it's a no-op.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: > then I think this could be done more simply with one of the `Arrays.copyOfRange` methods?

No, these are for arrays whose component type is statically known.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java [hunk 4]
```diff
@@ -102,6 +85,33 @@ private static boolean matches(@Nonnull Class<?>[] parameterTypes, @Nonnull Obje
...
+            if (arrayLength == 1 && parameterTypes[fixedLen].isInstance(parameters[fixedLen])) {
+                // not a varargs call
+                return parameters;
+            } else {
+                Object array = Array.newInstance(componentType, arrayLength);
+                for (int i = 0; i < arrayLength; i++) {
+                    Array.set(array, i, parameters[fixedLen + i]);
+                }
+                Object[] parameters2 = new Object[fixedLen + 1];
+                System.arraycopy(parameters, 0, parameters2, 0, fixedLen);
```
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: Super nitpicky 🐜 - can create and copy parameters2 with a single copyOfRange call (null will be placed in fixedLen until set).

**Comment by svanoort** [comment]
🐝 - the method refactor makes it isn't doing somewhat unexpected things with inputs.

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java [hunk 5]
```diff
@@ -471,6 +475,14 @@ public static int len(String... vals) {
             return vals.length;
         }
         @Whitelisted
+        public static int sum(int... numbers) {
```
Comment (svanoort) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java]: 😄  for seeing this tested.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 4 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: Yeah. Could.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java (hunk 2 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/GroovyCallSiteSelector.java]: Object is a type... though the primitives issues break this so no point now.

**Comment by recena** [comment]
:bee:
